File: icsopenapi/easyopen-sdk/sdk-csharp/EasyopenSDKCSharp/Request/BaseRequest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using EasyopenSDKCSharp.Common;
using EasyopenSDKCSharp.Response;
using EasyopenSDKCSharp.Utility;

namespace EasyopenSDKCSharp.Request
{
    /// <summary>
    /// 接口请求对象，新建的Request要继承这个类
    /// </summary>
    /// <typeparam name="T">对应的Response对象</typeparam>
    public abstract class BaseRequest<T>
    {
        private static string FORMAT_TYPE = SdkConfig.FORMAT_TYPE;
        private static string TIMESTAMP_PATTERN = SdkConfig.TIMESTAMP_PATTERN;
        private static RequestMethod METHOD_POST = RequestMethod.POST;


        private string name;
        /// <summary>
        /// 接口名
        /// </summary>
        public string Name
        {
            get { return name; }
            set { name = value; }
        }


        private string version;
        /// <summary>
        /// 版本号
        /// </summary>
        public string Version
        {
            get { return version; }
            set { version = value; }
        }

        private object data;
        /// <summary>
        /// 设置参数
        /// </summary>
        public object Data
        {
            get { return data; }
            set { data = value; }
        }
        /// <summary>
        /// 设置参数，同Data
        /// </summary>
        [IgnoreSign]
        public object Param
        {
            get { return data; }
            set { data = value; }
        }

        /// <summary>
        /// 时间戳
        /// </summary>
        private string timestamp = DateTime.Now.ToString(TIMESTAMP_PATTERN);
        public string Timestamp
        {
            get { return timestamp; }
            set { timestamp = value; }
        }

        private string access_token = string.Empty;
        public string Access_token
        {
            get { return access_token; }
            set { access_token = value; }
        }

        private string format = SdkConfig.FORMAT_TYPE;
        /// <summary>
        /// 传输格式，默认json
        /// </summary>
        public string Format
        {
            get { return format; }
            set { format = value; }
        }



        private RequestMethod requestMethod = METHOD_POST;
        /// <summary>
        /// 请求方式，默认POST
        /// </summary>
        [IgnoreSign]
        public RequestMethod RequestMethod
        {
            get { return requestMethod; }
            set { requestMethod = value; }
        }

        private List<UploadFile> files;
        /// <summary>
        /// 上传文件
        /// </summary>
        [IgnoreSign]
        public List<UploadFile> Files
        {
            get { return files; }
            set { files = value; }
        }



        public BaseRequest()
        {
            this.name = this.GetName();
            this.version = this.GetVersion();
        }

        public BaseRequest(string name, string version)
        {
            this.name = name;
            this.version = version;
        }

        /// <summary>
        /// 返回接口名
        /// </summary>
        /// <returns></returns>
        /// 
        public abstract string GetName();

        /// <summary>
        /// 返回版本号
        /// </summary>
        /// <returns></returns>
        public virtual string GetVersion()
        {
            return SdkConfig.DEFAULT_VERSION;
        }

        /// <summary>
        /// 创建请求表单
        /// </summary>
        /// <returns></returns>
        public RequestForm CreateRequestForm()
        {
            Dictionary<string, object> dict = ClassUtil.ConvertObjectToDictionary(this);
            RequestForm requestForm = new RequestForm(dict);
            requestForm.Files = this.files;
            return requestForm;
        }


    }
}

File: icsopenapi/easyopen-sdk/sdk-java/src/main/java/com/easyopen/sdk/request/BaseRequest.java
package com.easyopen.sdk.request;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.alibaba.fastjson.annotation.JSONField;
import com.easyopen.sdk.common.RequestForm;
import com.easyopen.sdk.common.RequestMethod;
import com.easyopen.sdk.common.UploadFile;
import com.easyopen.sdk.common.SdkConfig;
import com.easyopen.sdk.response.BaseResponse;
import com.easyopen.sdk.util.ClassUtil;

/**
 * 请求对象父类，后续请求对象都要继承这个类
 * @param <T> 对应的Response对象
 */
public abstract class BaseRequest<T extends BaseResponse<?>> {

    private static String FORMAT_TYPE = SdkConfig.FORMAT_TYPE;
    private static String TIMESTAMP_PATTERN = SdkConfig.TIMESTAMP_PATTERN;
    private static final String DEFAULT_ACCESS_TOKEN = "";

    private String name;
    private String version;
    private Object data;
    private String timestamp = new SimpleDateFormat(TIMESTAMP_PATTERN).format(new Date());
    private String access_token = DEFAULT_ACCESS_TOKEN;
    private String format = FORMAT_TYPE;

    /** 上传文件 */
    @JSONField(serialize = false)
    private List<UploadFile> files;

    @JSONField(serialize = false)
    private RequestMethod requestMethod = RequestMethod.POST;

    @JSONField(serialize = false)
    private Class<T> responseClass;

    @JSONField(serialize = false)
    public abstract String name();

    @SuppressWarnings("unchecked")
    public BaseRequest() {
        this.name = this.name();
        this.version = this.version();

        this.responseClass = (Class<T>) ClassUtil.getSuperClassGenricType(this.getClass(), 0);
    }

    @JSONField(serialize = false)
    protected String version() {
        return SdkConfig.DEFAULT_VERSION;
    }

    public RequestForm createRequestForm() {
        String json = JSON.toJSONString(this);
        JSONObject map = JSON.parseObject(json);
        RequestForm requestForm = new RequestForm(map);
        requestForm.setFiles(this.files);
        return requestForm;
    }

    public void setParam(Object param) {
        this.setData(param);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public Object getData() {
        return data;
    }

    /** 等同setParam() */
    public void setData(Object data) {
        this.data = data;
    }

    public String getAccess_token() {
        return access_token;
    }

    public void setAccess_token(String access_token) {
        this.access_token = access_token;
    }

    public String getFormat() {
        return format;
    }

    public void setFormat(String format) {
        this.format = format;
    }

    public RequestMethod getRequestMethod() {
        return requestMethod;
    }

    public void setRequestMethod(RequestMethod requestMethod) {
        this.requestMethod = requestMethod;
    }

    public String getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(String timestamp) {
        this.timestamp = timestamp;
    }

    public Class<T> getResponseClass() {
        return responseClass;
    }

    public List<UploadFile> getFiles() {
        return files;
    }

    /**
     * 添加上传文件
     * @param files
     */
    public void setFiles(List<UploadFile> files) {
        this.files = files;
    }
}
